### DISCUSSION TITLE
Client not receiving emits that are prompted by other emits

### DISCUSSION THREAD

**Discussion by Christiaan-Doubell** [miguelgrinberg/python-socketio#899] [Q&A] [answered]
Good day everyone

I have been having issues with using the python-socketio library to forward a message from one socket to another and would really appreciate your help to solve it.

I am trying to get a setup working where I have three python scripts - a client, a server, and a "gateway". The gateway is like a broker between the client and the server - it should connect with both the client and server on either side, and pass messages coming from the one to the other. The idea is that the gateway will later have to decide where to route requests to, when there are multiple clients and servers connected to it. 

### The issue
I want to do the following (illustrated in the picture below, with the blue indicating the role in relation to the other component):
![image](https://user-images.githubusercontent.com/102299167/160230274-40032176-27da-4b77-8106-afc07a797ce8.png)1. Emit an event (_from_server_) from the server to the gateway
2. The "gateway" receives the event and emits a subsequent event (_from_gateway_) to all the clients through the event-handler
3. The client receives the event and prints a message accordingly.

The issue comes in with the third point - everything happens as it should, until the client needs to receive the event.

### Code
I put below the code that I think might be relevant to this issue (additional to this I also perform the connect, disconnect, connect_error, etc. functions that are normally used when setting up the client-server connections).

**Gateway**
```
#Server socket setup
client_socket = socketio.Server(async_mode='eventlet', ping_interval=2, ping_timeout=10, cors_allowed_origins="*", logger=True, engineio_logger=True)
app = socketio.WSGIApp(client_socket)
client_address = 'http://localhost:5000'

#Client socket setup
server_socket = socketio.Client(logger=True, engineio_logger=True)

#-----------------------EVENTS----------------------------
#----------------------------------------------------------
#   Gateway-Client event-handlers
#------------------------------------------------
@server_socket.on('from_server')
def Handle_Client_Msg(msg):
    print('\n[GATEWAY] Received message from SERVER\n')    
    client_socket.emit('from_gateway', msg)
    print('\n[GATEWAY] Sent message to CLIENT\n')

#---------------------MAIN PROGRAM------------------
#---------------------------------------------------
#Main function
if __name__ == '__main__':
    dti_socket.start_background_task(connect_to_server, server_socket , server_address )
    listen_for_client('localhost', 5000, app)
```
where these two functions are declared to be:
```
def connect_to_server(socket, address):
    while socket.connected==False:
        try:
            socket.connect(address)
            socket.wait()
        except:
            print(f'\n[GATEWAY] No server at address: {address}\n[GATEWAY] Retrying...')

def listen_for_client(ip, port, web_app):
    eventlet.wsgi.server(eventlet.listen((ip, port)), web_app)
```

**Client**
```
sio_client = socketio.Client(logger=True, engineio_logger=True)
server_address = 'http://localhost:5000'

@sio_client.on('from_gateway')
def Handle_Gateway_Msg(msg):
    print('\n[UI] I received the message!\n')

if __name__ == '__main__':
    while True:
        connect_to_server(sio_client, server_address)
```
(The connect_to_server function is declared exactly the same in the client, as it is for the gateway.)

### Logs
Here is a snippet of the logs:

Gateway:
```
9KF6EIfydgBvqo8eAAAA: Sending packet PING data None
9KF6EIfydgBvqo8eAAAA: Received packet PONG data 
Received packet MESSAGE data 2["from_server","msg"]
Received event "from_server" [/]
[GATEWAY] Received message from SERVER
emitting event "from_gateway" to all [/]
9KF6EIfydgBvqo8eAAAA: Sending packet MESSAGE data 2["from_gateway","msg"]
[GATEWAY] Sent message to CLIENT
9KF6EIfydgBvqo8eAAAA: Sending packet PING data None
```
Client:
```
Received packet PING data
Sending packet PONG data 
Server has stopped communicating, aborting
Waiting for write loop task to end
Exiting write loop task
Engine.IO connection dropped
```

The client hangs for a while after sending the last PONG before it prints the rest of the logs, after which it disconnects and subsequently reconnects with a new sid.

### Testing the client's receiving ability
The strange thing is that if instead, I prompt the message from the Gateway function as follows:
```
def initiate():
    print('\n[GATEWAY] Starting message initiation\n')
    client_socket.sleep(5)
    client_socket.emit('from_gateway', 'msg')
    print('\n[GATEWAY] Sent message to CLIENT\n')

@client_socket.event
def connect(sid, environ):
    print('\n[GATEWAY] Connected to CLIENT\n')
    client_socket.start_background_task(initiate)
```

The client receives the message just fine.

Any help or advice would really be greatly appreciated.

**Comment by miguelgrinberg** [comment, marked as answer]
The Socket.IO client uses threads and low-level sockets. Both of these use blocking operations. In a multi-threaded application that is not a problem, but your gateway uses eventlet, which is an asynchronous framework, so those blocking calls are going to block everything.

Are you familiar with monkey patching? That might help. https://eventlet.net/doc/patching.html

**Reply by Christiaan-Doubell** [reply]
Thank you very much for your prompt response @miguelgrinberg. 
It works! 
By adding ```eventlet.monkey_patch()``` just after importing eventlet, the messages are received successfully on the client side.

Thank you so much!
